File: apps/app-frontend/src/helpers/threadrinth.ts
import { invoke } from '@tauri-apps/api/core'

/** Threadrinth release locations on GitHub. */
export const THREADRINTH_REPOSITORY_URL = 'https://github.com/Georgwav/Threadrinth'
export const THREADRINTH_RELEASES_URL = `${THREADRINTH_REPOSITORY_URL}/releases`
/** Updater manifest published with every release (same file the in-app updater reads). */
export const THREADRINTH_UPDATE_MANIFEST_URL = `${THREADRINTH_RELEASES_URL}/latest/download/latest.json`

export type WorldTransferMode = 'copy' | 'move'

/** Copies or moves a singleplayer world to another instance. Returns its folder name there. */
export async function transfer_world(
	fromInstance: string,
	world: string,
	toInstance: string,
	mode: WorldTransferMode,
): Promise<string> {
	return await invoke('plugin:threadrinth|transfer_world', {
		fromInstance,
		world,
		toInstance,
		mode,
	})
}

export type ServerPackReport = {
	mods_included: number
}

/** Include and exclude rules for the files the server pack export starts with. */
export type ServerPackSelection = {
	included: string[]
	excluded: string[]
}

/** The files a server needs: mods (minus client-only ones) and their configs. */
export async function server_pack_selection(instanceId: string): Promise<ServerPackSelection> {
	return await invoke('plugin:threadrinth|server_pack_selection', { instanceId })
}

/** Exports the selected files of an instance as a ready-to-run server zip. */
export async function export_server_pack(
	instanceId: string,
	exportLocation: string,
	included: string[],
	excluded: string[],
): Promise<ServerPackReport> {
	return await invoke('plugin:threadrinth|export_server_pack', {
		instanceId,
		exportLocation,
		included,
		excluded,
	})
}

File: apps/app/src/api/threadrinth.rs
//! Commands for Threadrinth's own features, kept apart from the upstream
//! plugins so Modrinth updates merge cleanly.

use crate::api::Result;
use std::path::PathBuf;
use tauri::Runtime;
use theseus::instance::{ServerPackReport, ServerPackSelection};
use theseus::world_transfer::WorldTransferMode;

pub fn init<R: Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("threadrinth")
        .invoke_handler(tauri::generate_handler![
            transfer_world,
            server_pack_selection,
            export_server_pack,
        ])
        .build()
}

/// Copies or moves a singleplayer world to another instance. Returns the
/// world's folder name there.
#[tauri::command]
pub async fn transfer_world(
    from_instance: &str,
    world: &str,
    to_instance: &str,
    mode: WorldTransferMode,
) -> Result<String> {
    Ok(theseus::world_transfer::transfer_world(
        from_instance,
        world,
        to_instance,
        mode,
    )
    .await?)
}

/// What the server pack export screen selects when it opens.
#[tauri::command]
pub async fn server_pack_selection(
    instance_id: &str,
) -> Result<ServerPackSelection> {
    Ok(theseus::instance::server_pack_selection(instance_id).await?)
}

/// Exports the selected files of an instance as a ready-to-run server zip.
#[tauri::command]
pub async fn export_server_pack(
    instance_id: &str,
    export_location: PathBuf,
    included: Vec<String>,
    excluded: Vec<String>,
) -> Result<ServerPackReport> {
    Ok(theseus::instance::export_server_pack(
        instance_id,
        export_location,
        included,
        excluded,
    )
    .await?)
}
